### PR TITLE
feat: better messaging for the user when having non sideloadable releases

### PR DESCRIPTION
**Pull request by erickzanardo**
## Description

When there were a release where one of the platforms were could not be sideloaded, that platform would simply not be show in the list.

This PR implements better messaging to the user regarding the non sideloadable platforms:
 - When no `platform`  is provided, we warn that a platform can't be side loaded
 - If the user provided a `platform`, and that platform can't be side loaded, we error instead

Fixes #2277 2154

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore
- [ ] 🧪 Tests

### CODE REVIEW

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 1]
```diff
@@ -114,15 +114,18 @@ class PreviewCommand extends ShorebirdCommand {
...
-      appId: appId,
-      sideloadableOnly: true,
-    );
+    final (allReleases, sideloadableReleases) = await (
+      codePushClientWrapper.getReleases(appId: appId),
+      codePushClientWrapper.getReleases(
+        appId: appId,
+        sideloadableOnly: true,
+      )
+    ).wait;
```
Comment (bryanoltman) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]: Why not just make a single request and filter for `isSideloadable`? That seems like it'd be a lot more efficient.

Comment (bryanoltman) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart]: Explained on VC. I'd leave a comment here to explain why we're doing this.

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 2]
```diff
@@ -143,12 +147,51 @@ class PreviewCommand extends ShorebirdCommand {
       return ExitCode.software.code;
     }
 
+    final completeRelease = allReleases.firstWhereOrNull(
```
Comment (bryanoltman) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]: The naming here is a bit confusing. What makes this release "complete"?

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 3]
```diff
@@ -143,12 +147,51 @@ class PreviewCommand extends ShorebirdCommand {
...
+          // previewable platform, we early exit to avoid duplicated warnings/errors.
+          if (results['platform'] == platform.name) {
+            logger.err(message);
+            return ExitCode.software.code;
+          } else {
+            logger.warn(message);
+          }
+        }
+      }
+    }
```
Comment (bryanoltman) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]: It might be good to extract this to a function. It's a bit hard to follow and it's not clear what exactly is being done here without a deeper reading

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 2 above)
Comment (bryanoltman) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart]: This release will be the same as `release` above, but will potentially contain more information if `release` contains non-previewable platforms (eg, an unsigned iOS release).

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 4]
```diff
@@ -143,12 +147,51 @@ class PreviewCommand extends ShorebirdCommand {
       return ExitCode.software.code;
     }
 
+    final completeRelease = allReleases.firstWhereOrNull(
+      (r) => r.version == releaseVersion,
+    );
+    if (completeRelease != null) {
```
Comment (bryanoltman) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]: I don't think that this is necessary, the set of releases with previewable artifacts should be a strict subset of the set of all releases

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 5]
```diff
@@ -143,12 +147,51 @@ class PreviewCommand extends ShorebirdCommand {
...
+          // If the user explicitly specified a platform and it matches a non
+          // previewable platform, we early exit to avoid duplicated warnings/errors.
+          if (results['platform'] == platform.name) {
+            logger.err(message);
+            return ExitCode.software.code;
+          } else {
+            logger.warn(message);
+          }
+        }
+      }
```
Comment (bryanoltman) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]: Breaking this out into a separate function would help with readability

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 6]
```diff
@@ -25,6 +25,8 @@ import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
 import 'package:yaml/yaml.dart';
 import 'package:yaml_edit/yaml_edit.dart';
 
+import '../third_party/flutter_tools/lib/flutter_tools.dart';
```
Comment (bryanoltman) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]:
```suggestion
import 'package:yaml_edit/yaml_edit.dart';
import '../third_party/flutter_tools/lib/flutter_tools.dart';
```

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 7]
```diff
@@ -77,6 +79,30 @@ class PreviewCommand extends ShorebirdCommand {
   @override
   String get description => 'Preview a specific release on a device.';
 
+  void _assertPreviwableReleases({
```
Comment (bryanoltman) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart:89]: I'd leave a comment explaining what we're asserting

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 8]
```diff
@@ -77,6 +78,38 @@ class PreviewCommand extends ShorebirdCommand {
...
 
+  /// Given two [Release]s, one with all the platforms, previewable or not,
+  /// and one with only the previewable platforms, this method will check for
+  /// platforms that are not previewable.
+  ///
+  /// If any is found, we will:
+  /// - Warn the user about the platform if that platform is not the one
+  /// the user provide on --platform
+  /// - Error out if the user provided the platform that is not previewable.
+  void _assertPreviwableReleases({
```
Comment (bryanoltman) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]:
```suggestion
  ///
  /// If any non-previewable platforms are found, we will:
  /// - Warn the user about the platform if that platform is not the one
  ///    the user specified with `--platform`.
  /// - Error out if the user specified a platform that is not previewable.
  void _assertPreviwableReleases({
```

**Comment by erickzanardo** [comment]
@bryanoltman Here are the outputs of the possible scenarios, in my test the release 1.0.0+1 has android and ios, only android is previewable

When not informing anything, it will ask for the release, warn that iOS is not previewable, and go straight to android
```
$ shorebird preview
✓ Fetching releases (0.4s)
✓ Fetching releases (0.4s)
Which release would you like to preview?
Which release would you like to preview? 1.0.0+1
[WARN] iOS is not previewable and can't be used.
✓ Fetching aab artifact (0.2s)
✓ Downloading release (0.2s)
✓ Using production track (0.2s)
✓ Extracting metadata (0.2s)
✓ Building apks (2.3s)
...
```

When passing `--platform ios`, it will break:

```
$ shorebird preview --platform ios
✓ Fetching releases (0.4s)
✓ Fetching releases (0.4s)
Which release would you like to preview?
Which release would you like to preview? 1.0.0+1
iOS is not previewable and can't be used.

If you aren't sure why this command failed, re-run with the --verbose flag to see more information.

You can also file an issue if you think this is a bug. Please include the following log file in your report:
/Users/erick/Library/Application Support/shorebird/logs/1718914944651_shorebird.log
```

And if I try preview choosing or passing as an argument of a release version, all works as it used to work before.

**Comment by bryanoltman** [comment]
I worry "[WARN] iOS is not previewable and can't be used." makes it sound like iOS as a platform is entirely not previewable, which obviously isn't the case. Maybe "The iOS artifact for this release is not previewable"?

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 9]
```diff
@@ -1267,6 +1278,376 @@ channel: ${DeploymentTrack.staging.channel}
           ).called(1);
         });
       });
+
+      group('when the android release is not sideloadable', () {
+        setUp(() {
+          final unfilteredRelease = MockRelease();
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]: I don't think it's clear what `unfiltered` means here

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 10]
```diff
@@ -1267,6 +1278,376 @@ channel: ${DeploymentTrack.staging.channel}
...
+          verify(
+            () => logger.err(
+              '''${ReleasePlatform.ios.displayName} is not previewable and can't be used.''',
+            ),
+          ).called(1);
+        });
+      });
+    });
+
+    group('when no platform is informed', () {
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]: When no platform is specified, I think?

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 11]
```diff
@@ -1267,6 +1278,376 @@ channel: ${DeploymentTrack.staging.channel}
...
+            when(
+              () => codePushClientWrapper.getReleases(
+                appId: any(named: 'appId'),
+              ),
+            ).thenAnswer((_) async => [unfilteredRelease]);
+          });
+
+          test('warns about the platform and goes directly to iOS', () async {
+            final exitCode = await runWithOverrides(command.run);
+            expect(exitCode, equals(ExitCode.success.code));
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]: I don't know that we need to warn about iOS if the user specified that they want to preview Android.

Comment (bryanoltman) [reply packages/shorebird_cli/test/src/commands/preview_command_test.dart]: Also, should the test name be "goes directly to android"?

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 12]
```diff
@@ -1267,6 +1278,376 @@ channel: ${DeploymentTrack.staging.channel}
...
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart:1533]: I don't think we need to warn about Android if the user specified that they want to preview iOS

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 13]
```diff
@@ -1267,6 +1278,376 @@ channel: ${DeploymentTrack.staging.channel}
...
+            deviceId: any(named: 'deviceId'),
+          ),
+        ).thenAnswer((_) async => ExitCode.success.code);
+        when(() => release.platformStatuses).thenReturn({
+          ReleasePlatform.android: ReleaseStatus.active,
+          ReleasePlatform.ios: ReleaseStatus.active,
+        });
+
+        iosReleaseArtifact = MockReleaseArtifact();
+        androidReleaseArtifact = MockReleaseArtifact();
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]: Prefer to instantiate mocks in a single block (these should be moved up with the other mock constructors)

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 14]
```diff
@@ -54,6 +56,101 @@ void main() {
...
+    String apksPath() {
+      final platformName = ReleasePlatform.android.name;
+      return p.join(
+        previewDirectory.path,
+        '${platformName}_${releaseVersion}_$androidArtifactId.apks',
+      );
+    }
+
+    File setupIOSShorebirdYaml() => File(
+          p.join(
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart:84]: Naming consistency nit: `setupIOSShorebirdYaml` vs `createAndroidShorebirdYaml`. These should line up.

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart (hunk 9 above)
Comment (erickzanardo) [reply packages/shorebird_cli/test/src/commands/preview_command_test.dart]: Renamed to `releatWithAllPlatforms` to match the name using in the command code

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart (hunk 10 above)
Comment (erickzanardo) [reply packages/shorebird_cli/test/src/commands/preview_command_test.dart]: yes! it will be a while for me `informed` in english doesnt work like it works in portuguese haha

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart (hunk 12 above)
Comment (erickzanardo) [reply packages/shorebird_cli/test/src/commands/preview_command_test.dart:1533]: Hum, that is fair! I will change the test and implement this additional case

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart (hunk 11 above)
Comment (erickzanardo) [reply packages/shorebird_cli/test/src/commands/preview_command_test.dart]: ah, this message is totally wrong, it was a remain from early tests that I was doing and we changed the approach since then.

I rewrote the message

**Comment by erickzanardo** [comment]
> I worry "[WARN] iOS is not previewable and can't be used." makes it sound like iOS as a platform is entirely not previewable, which obviously isn't the case. Maybe "The iOS artifact for this release is not previewable"?

Fair! I changed to the proposed message, it sounds better indeed!

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 15]
```diff
@@ -1267,6 +1278,381 @@ channel: ${DeploymentTrack.staging.channel}
...
+      }
+
+      setUp(() {
+        adb = MockAdb();
+        bundletool = MockBundleTool();
+        process = MockProcess();
+        devicectl = MockDevicectl();
+        iosDeploy = MockIOSDeploy();
+
+        iosReleaseArtifact = MockReleaseArtifact();
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]:
```suggestion
        iosDeploy = MockIOSDeploy();
        iosReleaseArtifact = MockReleaseArtifact();
```

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 16]
```diff
@@ -1267,6 +1278,381 @@ channel: ${DeploymentTrack.staging.channel}
...
+            releaseId: any(named: 'releaseId'),
+            arch: any(named: 'arch'),
+            platform: ReleasePlatform.android,
+          ),
+        ).thenAnswer((_) async => androidReleaseArtifact);
+
+        when(() => platform.isMacOS).thenReturn(true);
+      });
+
+      group('when the user choose ios', () {
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]:
```suggestion
      group('when the user chooses ios at the prompt', () {
```

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 17]
```diff
@@ -1267,6 +1278,381 @@ channel: ${DeploymentTrack.staging.channel}
...
+            verify(
+              () => logger.warn(
+                '''The ${ReleasePlatform.android.displayName} artifact for this release is not previewable.''',
+              ),
+            ).called(1);
+          });
+        });
+      });
+
+      group('when the user choose android', () {
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]:
```suggestion
      group('when the user chooses android at the prompt', () {
```

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 18]
```diff
@@ -1267,6 +1278,381 @@ channel: ${DeploymentTrack.staging.channel}
           ).called(1);
         });
       });
+
+      group('when the android release is not sideloadable', () {
+        setUp(() {
+          final releatWithAllPlatforms = MockRelease();
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]:
```suggestion
          final releaseWithAllPlatforms = MockRelease();
```

Comment (bryanoltman) [reply packages/shorebird_cli/test/src/commands/preview_command_test.dart]: Will need to be renamed throughout

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 19]
```diff
@@ -1267,6 +1278,381 @@ channel: ${DeploymentTrack.staging.channel}
...
+          ).thenAnswer((_) async => [release]);
+          when(
+            () => codePushClientWrapper.getReleases(
+              appId: any(named: 'appId'),
+            ),
+          ).thenAnswer((_) async => [releatWithAllPlatforms]);
+        });
+
+        test(
+          'does not warns since the user explicitly asked for iOS',
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]:
```suggestion
          'does not warn since the user explicitly asked for iOS',
```

File: packages/shorebird_cli/test/src/commands/preview_command_test.dart [hunk 20]
```diff
@@ -1267,6 +1278,381 @@ channel: ${DeploymentTrack.staging.channel}
...
+                '''The ${ReleasePlatform.android.displayName} artifact for this release is not previewable.''',
+              ),
+            );
+          },
+        );
+      });
+
+      group('when the ios release is not sideloadable', () {
+        setUp(() {
+          final releatWithAllPlatforms = MockRelease();
```
Comment (bryanoltman) [inline packages/shorebird_cli/test/src/commands/preview_command_test.dart]:
```suggestion
          final releaseWithAllPlatforms = MockRelease();
```

Same here

**Review by bryanoltman — APPROVED** [review]
